### PR TITLE
Do not be case-sensitive when checking http headers

### PR DESCRIPTION
**Pull request by ghost**
This is affecting policies evaluation in both the S3 and IAM services, when there is a condition on the IP.
No impact for S3C because the header is already properly formatted, but not in Zenko.

Issue: ARSN-453

### CODE REVIEW

**Review by KillianG — APPROVED** [review]

**Review by francoisferrand — DISMISSED** [review]
do we really need the extra conversion to lowercase, can't we just fix the 'X-Forwarded-For' where it is defined instead?

This adds a (small) overhead to every check, slightly increasing latency... and small things add up, over time...

File: lib/policyEvaluator/requestUtils.ts [hunk 1]
```diff
@@ -30,7 +30,10 @@ export function getClientIp(request: IncomingMessage, s3config?: S3Config): stri
          * which header to be used to extract client IP
          */
         if (ipCheck.ipMatchCidrList(trustedProxyCIDRs, clientIp)) {
-            const ipFromHeader = request.headers[extractClientIPFromHeader]?.toString();
+            // Request headers in nodejs are lower-cased, so we should not
+            // be case-sentive when looking for the header, as http headers
+            // are case-insensitive.
+            const ipFromHeader = request.headers[extractClientIPFromHeader.toLowerCase()]?.toString();
```
Comment (francoisferrand) [inline lib/policyEvaluator/requestUtils.ts:32]: looking a few lines above, we already retrieve the x-forwarded-for header:

```
const clientIp = (requestConfig ? remoteAddress : request.headers['x-forwarded-for'] || remoteAddress)?.toString() ?? '';
```

this seems inconsistent, please take the chance to review and clean this...
Esp. if `requestConfig` is not provided, we kind of bindly trust the `x-forwarded-for` header : is this really expected?

Comment (ghost) [reply lib/policyEvaluator/requestUtils.ts:32]: > do we really need the extra conversion to lowercase, can't we just fix the 'X-Forwarded-For' where it is defined instead?

We could do it but the headers are case-insensitive anyway, so the current code is not following the standard and we might have more issues in the future...

Comment (ghost) [reply lib/policyEvaluator/requestUtils.ts:32]: > looking a few lines above, we already retrieve the x-forwarded-for header:
> const clientIp = (requestConfig ? remoteAddress : request.headers['x-forwarded-for'] || remoteAddress)?.toString() ?? '';
this seems inconsistent, please take the chance to review and clean this...
Esp. if requestConfig is not provided, we kind of bindly trust the x-forwarded-for header : is this really expected?

I am not aware of any deployment today that does not set the `requestConfig`, but this header retrieval was introduced in https://scality.atlassian.net/browse/ARSN-57, so here: https://github.com/scality/Arsenal/pull/1693/files#diff-ca8ab692737b69e1ab977a7a6104be671972410cafadea45b119e8259070ad1bR11

If the question is whether we should trust the header blindly or not: as long as the proxy is not trusted we should not trust the header. I will remove this logic to be on the safe side.

Previously we were not using IP conditions so it was only about displaying the info in logs, hence the previous change.

Comment (ghost) [reply lib/policyEvaluator/requestUtils.ts:32]: > extractClientIPFromHeader probably does not work if not equal to x-forwarded-for

Please elaborate what you mean by "does not work": it's a string field, that can be any header: e.g., we can have a custom header set by nginx and use it.

Comment (francoisferrand) [reply lib/policyEvaluator/requestUtils.ts:32]: > extractClientIPFromHeader probably does not work if not equal to x-forwarded-for
> > Please elaborate what you mean by "does not work": it's a string field, that can be any header: e.g., we can have a custom header set by nginx and use it.

"does not work" was too quick, I missed the condition on `requestConfig`

**Review by francoisferrand — APPROVED** [review]

**Pull request merged**